src/core/or/connection_or.c

```diff
@@ -726,6 +726,18 @@ connection_or_finished_flushing(or_connection_t *conn)
 
   switch (conn->base_.state) {
     case OR_CONN_STATE_PROXY_HANDSHAKING:
+      /* PROXY_HAPROXY gets connected by receiving an ack. */
+      if (conn->proxy_type == PROXY_HAPROXY) {
+        tor_assert(TO_CONN(conn)->proxy_state == PROXY_HAPROXY_WANT_ACK);
+        TO_CONN(conn)->proxy_state = PROXY_CONNECTED;
+
```

Comment:
Receiving ACK means all the data is flushed out of the buffer. When the ACK is received, `connection_or_finished_flushing` will be triggered.

According to the haproxy spec, it can do a tls handshake immediately after sending the PROXY header. But it seems like if we do the tls handshake immediately (without waiting for an ack first) something will crash (I don't know why, it may be because there's something in the buffer left).

Reply:
Hm.  In that case, let's rename this to WAIT_FOR_FLUSH or something -- an "ACK" usually means an explicit message from somebody else.